### ISSUE TITLE
Autokey just stops expanding, narrowed it down to UPPER case issue

### ISSUE THREAD

**Issue by nick-s-b** [autokey/autokey#95]
## Classification:

Bug

## Reproducibility:

Almost every day after few hours of use.

## Summary

I have noticed that Autokey just stops expanding after some time has passed (usually hours). It has baffled me for weeks and I've basically resorted to restarting it once it stopped working. Finally today, I spent some time trying to track down the issue and have started Autokey from the command line with verbose argument.

Every now and then, I'd try Autokey and see if it still expanded. After a while, Autokey stopped working and I checked the output. This is what I saw when I tried to expand the "expansion1" phrase that I setup earlier:

````
2017-08-13 10:21:29,843 DEBUG - service - Raw key: '<backspace>', modifiers: [], Key: <backspace>
2017-08-13 10:21:29,843 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:29,843 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,085 DEBUG - service - Raw key: 'a', modifiers: [], Key: E
2017-08-13 10:21:30,085 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,085 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,085 DEBUG - service - Input stack at end of handle_keypress: deque(['E'], maxlen=150)
2017-08-13 10:21:30,268 DEBUG - service - Raw key: 'p', modifiers: [], Key: X
2017-08-13 10:21:30,268 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,268 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,269 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X'], maxlen=150)
2017-08-13 10:21:30,455 DEBUG - service - Raw key: 'p', modifiers: [], Key: P
2017-08-13 10:21:30,455 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,455 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,456 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P'], maxlen=150)
2017-08-13 10:21:30,652 DEBUG - service - Raw key: 'l', modifiers: [], Key: A
2017-08-13 10:21:30,652 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,652 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,653 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A'], maxlen=150)
2017-08-13 10:21:30,782 DEBUG - service - Raw key: 'e', modifiers: [], Key: N
2017-08-13 10:21:30,782 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,782 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,783 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N'], maxlen=150)
2017-08-13 10:21:30,985 DEBUG - service - Raw key: 'h', modifiers: [], Key: S
2017-08-13 10:21:30,985 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:30,985 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:30,986 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S'], maxlen=150)
2017-08-13 10:21:31,114 DEBUG - service - Raw key: 'a', modifiers: [], Key: I
2017-08-13 10:21:31,115 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:31,115 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:31,115 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S', 'I'], maxlen=150)
2017-08-13 10:21:31,214 DEBUG - service - Raw key: 't', modifiers: [], Key: O
2017-08-13 10:21:31,214 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:31,214 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:31,214 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S', 'I', 'O'], maxlen=150)
2017-08-13 10:21:31,330 DEBUG - service - Raw key: 'e', modifiers: [], Key: N
2017-08-13 10:21:31,331 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:31,331 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:31,331 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S', 'I', 'O', 'N'], maxlen=150)
2017-08-13 10:21:31,520 DEBUG - service - Raw key: 'r', modifiers: [], Key: 1
2017-08-13 10:21:31,520 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:31,520 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:31,520 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S', 'I', 'O', 'N', '1'], maxlen=150)
2017-08-13 10:21:31,641 DEBUG - service - Raw key: ' ', modifiers: [], Key:  
2017-08-13 10:21:31,642 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
2017-08-13 10:21:31,642 DEBUG - service - No phrase/script matched hotkey
2017-08-13 10:21:31,642 DEBUG - service - Input stack at end of handle_keypress: deque(['E', 'X', 'P', 'A', 'N', 'S', 'I', 'O', 'N', '1', ' '], maxlen=150)
2017-08-13 10:21:31,995 DEBUG - iomediator - <ctrl> pressed
2017-08-13 10:21:32,151 DEBUG - service - Raw key: '<backspace>', modifiers: ['<ctrl>'], Key: <backspace>
2017-08-13 10:21:32,151 DEBUG - service - Window visible title: b'~/Documents/Text/test.txt \xe2\x80\xa2 - Sublime Text', Window class: 'subl3.Subl3'
````

Notice how all the keys are capitals? The reason why it stops expanding is because it thinks I typed "EXPANSION1" and not "expansion1". I don't have "ignore case" selected.

I was definitely typing in lower case. In fact, I have a CapsLock key disabled so it's impossible for me to type something in all upper case by accident. 

And notice how Autokey reports `Raw key: 'r', modifiers: [], Key: R` ? So no modifiers yet it converts it into upper case. When Autokey works properly, we see this for 'R': `Raw key: 'r', modifiers: ['<shift>'], Key: R`.
That's definitely a bug somewhere.

## Version

autokey-gtk 0.93.10

Installed via: AUR

Distro: Arch

## Workaround
Select ignore case for each abbreviation. But this might introduce some other issues (conflicts) depending on your abbreviations.

**EDIT:** It happened again and now when I tried typing the abbreviation with a Shift key pressed, it worked. Here's the log: `Raw key: 'r', modifiers: ['<shift>'], Key: r`. Some key/mouse event triggers Autokey's modifiers flag and Autokey forgets to clear it and flip it back.

**Comment by caleb** [comment]
Hmm, I've been experiencing similar symptoms. I wonder if it's the same thing. I'd like to figure this one out.

**Comment by caleb** [comment]
@nick-s-b 

Okay, I've found my problem and the verbose output looks like yours with the capital letters.

I switch my caps lock key to be another control key (via the Gnome tweak tool), and it appears that autokey, or something else is keeping track of my caps lock state when I hit my caps lock key even though on my system it doesn't actually enable caps lock.

If autokey stops working all I have to do is hit caps lock again to toggle this invisible caps lock state and it starts working again.

Again, I don't know if this is an issue with autokey or with something else in the system.

-Caleb

**Comment by nick-s-b** [comment]
@caleb Interesting! Thanks for looking into this. Actually, my Caps Lock key is completely disabled with Cinnamon/GNOME Keyboard settings. So I can't even engage it. The fault is with Autokey for sure... it's not properly keeping tracks of modifier keys.

**Comment by caleb** [comment]
So if autokey is working and you hit your caps lock key does autokey stop working for you?

If I get the time I will look into fixing this.

**Comment by nick-s-b** [comment]
@caleb Not sure... my CapsLock is non-existent and never registers as a keypress. If I hit it, it's registered at Ctrl. But I rarely hit that key anyway and Autokey fails to expand without me hitting it. No idea what's causing this bug at all. I looked at Autokey source but couldn't narrow it down. Please try to figure it out. It's extremely annoying :) Thanks.
PS: Autokey seems to be abandonware yet again :(

**Comment by caleb** [comment]
I've worked around this bug in my version of autokey:

```diff
diff --git a/lib/autokey/iomediator_constants.py b/lib/autokey/iomediator_constants.py
index b6ff9f0..3a720a3 100644
--- a/lib/autokey/iomediator_constants.py
+++ b/lib/autokey/iomediator_constants.py
@@ -3,5 +3,5 @@ X_RECORD_INTERFACE = "XRecord"
 KEY_SPLIT_RE = re.compile("(<[^<>]+>\+?)")
 
 from .iomediator_Key import Key
-MODIFIERS = [Key.CONTROL, Key.ALT, Key.ALT_GR, Key.SHIFT, Key.SUPER, Key.HYPER, Key.META, Key.CAPSLOCK, Key.NUMLOCK]
+MODIFIERS = [Key.CONTROL, Key.ALT, Key.ALT_GR, Key.SHIFT, Key.SUPER, Key.HYPER, Key.META, Key.NUMLOCK]
 HELD_MODIFIERS = [Key.CONTROL, Key.ALT, Key.SUPER, Key.SHIFT, Key.HYPER, Key.META]
```

I just removed the `Key.CAPSLOCK` from the `MODIFIERS` list, and it seems to work for me.

This isn't a solution, and I don't think in my case that autokey is to blame since it is receiving caps locks key press events even though I've turned the caps lock key into a control key.

Your problem seems different, but who knows...

**Comment by allanlaal** [comment]
same/similar issue in 0.90.4 - Autokey just stops expanding and needs to be restarted after X hours of use

toggling caps lock and writing stuff does not trigger this however

**Comment by avatar382** [comment]
I'm seeing the same behavior on my system (Kubuntu 18.04). Autokey will stop expanding phrases after I press Caps Lock, and it starts working again after I press Caps Lock a second time.

I have Caps Lock mapped to a control key via the KDE settings app (*Run* System Settings -> Input Devices -> Keyboard -> Advanced -> Caps Lock behavior -> *check on* Caps Lock is also a Ctrl)

@caleb, I'm not able to find the iomediator_constants.py file to modify it as you have. Here's the output I'm seeing for `dpkg -L autokey-gtk`:

```/usr
/usr/bin
/usr/bin/autokey-gtk
/usr/lib
/usr/lib/python2.7
/usr/lib/python2.7/dist-packages
/usr/lib/python2.7/dist-packages/autokey
/usr/lib/python2.7/dist-packages/autokey/gtkapp.py
/usr/lib/python2.7/dist-packages/autokey/gtkui
/usr/lib/python2.7/dist-packages/autokey/gtkui/__init__.py
/usr/lib/python2.7/dist-packages/autokey/gtkui/configwindow.py
/usr/lib/python2.7/dist-packages/autokey/gtkui/data
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/abbrsettings.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/autokey.svg
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/blankpage.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/detectdialog.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/folderpage.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/hotkeysettings.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/mainwindow.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/menus.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/phrasepage.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/recorddialog.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/renamedialog.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/scriptpage.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/settingsdialog.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/settingswidget.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/data/windowfiltersettings.xml
/usr/lib/python2.7/dist-packages/autokey/gtkui/dialogs.py
/usr/lib/python2.7/dist-packages/autokey/gtkui/notifier.py
/usr/lib/python2.7/dist-packages/autokey/gtkui/popupmenu.py
/usr/lib/python2.7/dist-packages/autokey/gtkui/settingsdialog.py
/usr/share
/usr/share/applications
/usr/share/applications/autokey-gtk.desktop
/usr/share/doc
/usr/share/doc/autokey-gtk
/usr/share/doc/autokey-gtk/copyright
/usr/share/man
/usr/share/man/man1
/usr/share/man/man1/autokey-gtk.1.gz
/usr/share/doc/autokey-gtk/changelog.Debian.gz
```

Do I need to change a different file? Thanks for reporting this, this issue was driving me bonkers!

**Comment by luziferius** [comment]
@avatar382 It is located in the Ubuntu package `autokey-common` (https://packages.ubuntu.com/bionic/all/autokey-common/filelist)
Just note that the version supplied by Ubuntu is about 6 years old. You’ll need to modify `/usr/lib/python2.7/dist-packages/autokey/iomediator.py` instead.

It looks like there is some common cause: A disabled caps lock key. Somehow AutoKey gets to know that a caps lock key is pressed and keeps tracking its state.

**Comment by sporkwitch** [comment]
@avatar382 please try using the current version (there's a PPA in the readme); this _should_ be fixed.

cc: @luziferius

**Comment by avatar382** [comment]
@sporkwitch I've added the `ppa:sporkwitch/autokey repo`, removed the old `autokey-gtk` and `autokey-common` packages, and then re-installed `autokey-gtk` package, but I'm still seeing the same behavior.

Is `0.94.1-0ubuntu3` the correct version?

**Comment by avatar382** [comment]
@luziferius Thanks! On my 18.04 system, the file was in `/usr/lib/python3/dist-packages/autokey/iomediator/_iomediator.py`.

I had to modify several lines of code, removing all references to Key.CAPS_LOCK, which has resolved this issue for me. It helps that I never use Caps Lock :)

**Comment by luziferius** [comment]
@sporkwitch I have not seen any commits that may have the effect of fixing this, as nothing in the modifier key handling has changed recently.
As far as I understand it, it is caused by users remapping their [CAPS LOCK] key to something else, like [Ctrl]. AutoKey still sees the Caps lock key event (although it should not, or should at least not react to it) and toggles its internal modifier tracking for the [Shift] modifier.
This, in turn, causes all key presses done by autokey to be case inverted.

**Comment by caleb** [comment]
Another solution is to turn off case sensitivity for your triggers. That also worked for me.

**Comment by sporkwitch** [comment]
@luziferius Apologies if i was mistaken, I thought i'd seen another note where the case-based issue was resolved; wasn't aware this was caps-lock specific.

**Comment by rattletat** [comment]
Any update on this? Having the same problem mapping escape to caps_lock using xcape.
Setting the ignore case option works, but isn't the best solution.
Especially when using a  `;` as prefix key and having to add `:` as a prefix too.

**Comment by luziferius** [comment]
Well, it took some time…
I think I got a fix for this, here: https://github.com/autokey/autokey/tree/bugfix/issue_291

To fix this issue, I added an option to the settings dialogue to disable the internal Capslock handling. If this is checked, Capslock is treated like any other regular key. See below the list for an explanation.

To try out the fix, do:

- Just in case, backup `~/.config/autokey/autokey.json`
- Quit any running AutoKey instance
- Clone the AutoKey git repository
- Run `git checkout bugfix/issue_291`. git should tell you `Switched to branch 'bugfix/issue_291'`
- Then follow https://github.com/autokey/autokey#zero-installation-method
- When Autokey started, open the settings/preferences dialogue and check the option `Disable handling of the Capslock key` and hit Save
- If the fix does not directly work, try restarting AutoKey. But in this case, please report that a restart was needed.


The reasoning behind the issue: AutoKey internally keeps track of the modifier key states. The API AutoKey uses (XRecord) does not apply the Capslock modifier internally, so AutoKey has to keep track and do it itself. In case of a remapped Capslock, AutoKey still sees the Capslock press and thinks that it has to invert the letter case for all read and written letters. So the internal state diverges from the system state, causing AutoKey to break.

**Comment by luziferius** [comment]
Just released 0.95.10, which contains a fix. You can now disable Capslock handling in the settings.
